packages/lesswrong/lib/sql/Type.ts

```diff
@@ -0,0 +1,255 @@
+      case String:
+        return typeof schema.foreignKey === "string"
+          ? new IdType(getCollection(schema.foreignKey as CollectionNameString))
+          : new StringType(typeof schema.max === "number" ? schema.max : undefined);
+      case Boolean:
+        return new BoolType();
+      case Date:
+        return new DateType();
+      case Number:
+        return new FloatType();
+      case "SimpleSchema.Integer":
+        return new IntType();
+      case Object: case GraphQLJSON:
+        return new JsonType();
+      case Array:
+        if (!indexSchema) {
+          throw new Error("No schema type provided for array member");
+        }
+        return new ArrayType(Type.fromSchema(fieldName + ".$", indexSchema));
+    }
+
+    if (schema.type instanceof SimpleSchema) {
+      return new JsonType();
+    }
+
+    throw new Error(`Unrecognized schema: ${JSON.stringify(schema)}`);
+  }
+}
+
+export class StringType extends Type {
+  constructor(private maxLength?: number) {
+    super();
+  }
+
+  toString() {
+    return this.maxLength === undefined ? "TEXT" : `VARCHAR(${this.maxLength})`;
+  }
+
+  getIndexCoalesceValue(): string | null {
+    return "''";
+  }
+}
+
+export class BoolType extends Type {
+  toString() {
+    return "BOOL";
+  }
+}
+
+export class IntType extends Type {
+  toString() {
+    return "INTEGER";
+  }
+}
+
+export class FloatType extends Type {
+  toString() {
+    return "REAL";
+  }
+}
+
+export class DateType extends Type {
+  toString() {
+    return "TIMESTAMPTZ";
+  }
+}
+
+export class JsonType extends Type {
+  toString() {
+    return "JSONB";
+  }
+}
+
+export class ArrayType extends Type {
+  subtype: Type;
+
+  constructor(subtype: Type) {
+    super();
+    this.subtype = subtype.toConcrete();
+  }
+
+  toString() {
+    return `${this.subtype.toString()}[]`;
+  }
+
+  isArray() {
+    return true;
+  }
+}
+
+/**
+ * IdType is a convinience type to automatically make sure Vulcan
+ * ID fields are stored correctly and efficiently
+ * By default, our IDs are 17 characters (ID_LENGTH), but some legacy
+ * data in Mongo uses ObjectId types which can be longer, so we add
+ * an extra 10 characters for safety. In the future, we may be able to
+ * reduce this after cleaning up the data, but it's too dangerous to do
+ * right now.
+ */
+export class IdType extends StringType {
+  constructor(private collection?: CollectionBase<any>) {
+    super(ID_LENGTH + 10);
+  }
+
+  getCollection() {
+    return this.collection;
+  }
+}
+
+/**
+ * Annotate a type as being non-nullable. Subtype may or may not be concrete.
+ */
+export class NotNullType extends Type {
+  constructor(private type: Type) {
+    super();
+  }
+
+  toString() {
+    return `${this.type.toString()} NOT NULL`;
+  }
+
+  toConcrete() {
+    return this.type.toConcrete();
+  }
+
+  isArray() {
+    return this.type.isArray();
+  }
+}
+
+const sqlEscape = (data: string): string => data.replace(/((?<!')('{2})*'(?!'))/g, "$1'");
```

Comment:
So, it turns out this is actually wrong. Originally, I forgot that escaping in SQL is `''` instead of `\'`, in which case this regex is correct because you only want to match on an odd number of backslashes, as an even number of backslashes is just a bunch of escaped backslashes. Now that I corrected it, we don't need this tricky logic anymore - we can just do a dumb replace of every `'` with `''`...